nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveStreaming.java

```diff
@@ -0,0 +1,657 @@
+                partitionColumnList = Collections.emptyList();
+            } else {
+                String[] partitionCols = partitionColumns.split(",");
+                partitionColumnList = new ArrayList<>(partitionCols.length);
+                for (String col : partitionCols) {
+                    partitionColumnList.add(col.trim());
+                }
+            }
+
+            // Store the original class loader, then explicitly set it to this class's classloader (for use by the Hive Metastore)
+            ClassLoader originalClassloader = Thread.currentThread().getContextClassLoader();
+            Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+
+            int recordCount = 0;
+            final List<HiveStreamingRecord> records = new LinkedList<>();
+
+            session.read(flowFile, in -> {
+
+                try (final DataFileStream<GenericRecord> reader = new DataFileStream<>(in, new GenericDatumReader<GenericRecord>())) {
+
+                    GenericRecord currRecord;
+                    while (reader.hasNext()) {
+                        currRecord = reader.next();
+                        List<String> partitionValues = new ArrayList<>();
+
+                        for (String partition : partitionColumnList) {
+                            Object partitionValue = currRecord.get(partition);
+                            if (partitionValue == null) {
+                                throw new IOException("Partition column '" + partition + "' not found in Avro record");
+                            }
+                            partitionValues.add(partitionValue.toString());
+                        }
+
+                        List<Schema.Field> fields = currRecord.getSchema().getFields();
+                        if (fields != null) {
+                            JSONObject obj = new JSONObject();
+                            for (Schema.Field field : fields) {
+                                String fieldName = field.name();
+                                // Skip fields that are partition columns, we extracted those values above to create an EndPoint
+                                if (!partitionColumnList.contains(fieldName)) {
+                                    Object value = currRecord.get(fieldName);
+                                    try {
+                                        obj.put(fieldName, value);
+                                    } catch (JSONException je) {
+                                        throw new IOException(je);
+                                    }
+                                }
+                            }
+                            records.add(new HiveStreamingRecord(partitionValues, obj));
+                        }
+                    }
+                }
+            });
+
+            // Write all records to Hive Streaming
+            for (HiveStreamingRecord record : records) {
```

Comment:
HiveWriter will batch up a number of records for efficiency, but you're right, here I'm waiting until all the records are in memory. There was some reason (probably incorrect) that I had it outside the loop, but if that reason was only to keep the recordCount (which would need to be final for the callback), I can just use an AtomicInteger instead.